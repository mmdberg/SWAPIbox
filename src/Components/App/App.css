.App {
/*  text-align: center;*/
/*  display: grid;
  grid-template-rows: 100px 200px 300px;
  grid-template-columns: 25% 65% 10%; */
}

h1 {
  text-shadow: #ecdd13 0px 0px 10px;
}

.header {
  grid-row: 1 / 2;
  grid-column: 2 / 3;
  background-color: chartreuse;
}

.logo {
  font-size: 2.5em;
  text-align: center;
}

.scroll {
  background-color: black;
  color: white;
  height: 300px;
  grid-row: 1 / 4;
  grid-column: 1 / 2;
}

.button-container {
  background-color: coral;
  grid-row: 2 / 3;
  grid-column: 2 / 4;
}



.card-container {
  background-color: aqua;
  height: 100px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}


