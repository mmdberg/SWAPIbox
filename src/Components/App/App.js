import React, { Component } from 'react';
import './App.css';
import { Scroll } from '../Scroll/index'
import { Favorites } from '../Favorites/index';
import { ButtonContainer } from '../ButtonContainer/index';
import { CardContainer } from '../CardContainer/index';

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      opening: {},
      cards: [],
      errorStatus: ''
    }
  }

  getOpening = async () => {
    try { let opening = await this.props.openingCall()
    this.setState({ opening }) } catch (err) {
      this.setState({errorStatus: 'Error loading data'})
    }
  }

  getCards = async (userInput) => {
    try { let cards = await this.props.buttonCall(userInput)
     this.setState({ cards }) 
    } catch (err) {
      this.setState({errorStatus: 'Error loading data'})
     }
  }

  componentDidMount() {
    this.getOpening();
  }

  render() {
    return (
      <div className="App">
        <header className='header'>
          <h1 className="logo">SWAPIbox</h1>
        </header>
        <Scroll opening={this.state.opening}/>
        <Favorites />
        <ButtonContainer getCards={this.getCards}/>
        <CardContainer cards={this.state.cards}/>
      </div>
    );
  }
}

export default App;
