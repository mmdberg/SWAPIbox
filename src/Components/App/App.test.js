import React from 'react';
import App from './App';
import { shallow } from 'enzyme';

describe('App', () => {

  it('matches the snapshot', () => {
    let wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should start with empty states', () => {
    let wrapper = shallow(<App />, {disableLifecycleMethods: true});
    let expectedState = {
      opening: {},
      cards: [],
      errorStatus: '',
      favorites: []
    };
    
    expect(wrapper.state()).toEqual(expectedState);
  });

  it('getOpening should update state with text, title and year', async () => {
    let expectedOpening = {
      text: "It is a period of civil war. Rebel spaceships.",
      title: "A New Hope",
      releaseYear: "1977"
    };
    let mockOpeningCall = jest.fn().mockImplementation(() => {
      return expectedOpening;
    });
    let wrapper = shallow(<App openingCall={mockOpeningCall}/>);

    await wrapper.instance().getOpening();
    expect(wrapper.state('opening')).toEqual(expectedOpening);

  });

  it('getOpening should set error state on error', async () => {
    let wrapper = shallow(<App />);

    window.fetch = jest.fn().mockImplementation(() => Promise.reject(
      status: 404
    ));
    const expected = 'Error loading data';

    await wrapper.instance().getOpening();
    expect(wrapper.state('errorStatus')).toEqual(expected);
  });

  it('getCards should update cards state', async () => {
    let expected = {
      class: "wheeled",
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30"
    };
    let mockButtonCall = jest.fn().mockImplementation(() => {
      return expected;
    });
    let wrapper = shallow(<App buttonCall={mockButtonCall}/>);

    await wrapper.instance().getCards('vehicles');
    expect(wrapper.state('cards')).toEqual(expected);
  });

  it('getCards should set error state on error', async () => {
    let wrapper = shallow(<App />);

    window.fetch = jest.fn().mockImplementation(() => Promise.reject(
      status: 404
    ));

    const expected = 'Error loading data';

    await wrapper.instance().getCards('vehicles');
    expect(wrapper.state('errorStatus')).toEqual(expected);
  });

  it('handleFavorites should add card to favorites if not in there', () => {
    let wrapper = shallow(<App />);
    let mockCard = {
      class: "wheeled",
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30"
    };

    wrapper.instance().handleFavorites(mockCard);
    expect(wrapper.state('favorites')).toEqual([mockCard]);
  });

  it('handleFavorites should remove card from favorites if there', () => {
    let wrapper = shallow(<App />);
    let mockCard = {
      class: "wheeled",
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30"
    };

    wrapper.setState({
      favorites: [mockCard]
    });
    wrapper.instance().handleFavorites(mockCard);
    expect(wrapper.state('favorites')).toEqual([]);
  });

  it('clearCards should reset cards and favorites in state', () => {
    let wrapper = shallow(<App />);
    let mockCard = {
      class: "wheeled",
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30"
    };

    wrapper.setState({
      cards: mockCard,
      favorites: mockCard
    })

    wrapper.instance().clearCards()
    expect(wrapper.state('cards')).toEqual([])
    expect(wrapper.state('favorites')).toEqual([])

  })

});
