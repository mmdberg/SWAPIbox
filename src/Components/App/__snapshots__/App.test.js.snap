// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="header"
>
          <Link
                    onClick={[Function]}
                    replace={false}
                    to="/"
          >
                    <img
                              alt="logo"
                              className="logo"
                              src="swapi-box-title.png"
                    />
          </Link>
</header>,
        <Route
          exact={true}
          path="/"
          render={[Function]}
/>,
        <Route
          path="/home/"
          render={[Function]}
/>,
        <Route
          exact={true}
          path="/favorites"
          render={[Function]}
/>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Link
            onClick={[Function]}
            replace={false}
            to="/"
>
            <img
                        alt="logo"
                        className="logo"
                        src="swapi-box-title.png"
            />
</Link>,
          "className": "header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <img
              alt="logo"
              className="logo"
              src="swapi-box-title.png"
/>,
            "onClick": [Function],
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "logo",
              "className": "logo",
              "src": "swapi-box-title.png",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": [Function],
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "path": "/home/",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/favorites",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="header"
>
            <Link
                        onClick={[Function]}
                        replace={false}
                        to="/"
            >
                        <img
                                    alt="logo"
                                    className="logo"
                                    src="swapi-box-title.png"
                        />
            </Link>
</header>,
          <Route
            exact={true}
            path="/"
            render={[Function]}
/>,
          <Route
            path="/home/"
            render={[Function]}
/>,
          <Route
            exact={true}
            path="/favorites"
            render={[Function]}
/>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Link
              onClick={[Function]}
              replace={false}
              to="/"
>
              <img
                            alt="logo"
                            className="logo"
                            src="swapi-box-title.png"
              />
</Link>,
            "className": "header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <img
                alt="logo"
                className="logo"
                src="swapi-box-title.png"
/>,
              "onClick": [Function],
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "logo",
                "className": "logo",
                "src": "swapi-box-title.png",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": [Function],
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "path": "/home/",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/favorites",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
